cudamapper/src/main.cpp

```diff
@@ -66,27 +66,101 @@ int main(int argc, char *argv[])
-    claragenomics::Matcher matcher(*index);
-    CGA_LOG_INFO("Finished matcher");
-    std::cerr << "Matcher execution time: " << std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - start_time).count() << "ms" << std::endl;
-
-    start_time = std::chrono::high_resolution_clock::now();
-    CGA_LOG_INFO("Started overlap detector");
-    auto overlapper = claragenomics::OverlapperTriggered();
-    auto overlaps = overlapper.get_overlaps(matcher.anchors(), *index);
-
-    CGA_LOG_INFO("Finished overlap detector");
-    std::cerr << "Overlap detection execution time: " << std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::high_resolution_clock::now() - start_time).count() << "ms" << std::endl;
-
-    overlapper.print_paf(overlaps);
-    return  0;
+
+    while(true) { // outer loop over query
+
+        //first generate a2a for query
+        std::vector<std::pair<std::uint64_t, std::uint64_t>> ranges;
+        std::pair<std::uint64_t, std::uint64_t> query_range {query_start, query_end};
+
+        ranges.push_back(query_range);
+
+        std::unique_ptr<claragenomics::Index> index = claragenomics::Index::create_index(input_filepath, k, w, ranges);
+
+        CGA_LOG_INFO("Created index");
+        std::cerr << "Index execution time: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::high_resolution_clock::now() - start_time).count() << "ms" << std::endl;
+
+        //auto num_reads = index.get()->number_of_reads();
+        auto match_point = 0; // all to all
+
+        start_time = std::chrono::high_resolution_clock::now();
+        CGA_LOG_INFO("Started matcher");
+        claragenomics::Matcher matcher(*index, match_point);
+        CGA_LOG_INFO("Finished matcher");
+        std::cerr << "Matcher execution time: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::high_resolution_clock::now() - start_time).count() << "ms" << std::endl;
+
+        start_time = std::chrono::high_resolution_clock::now();
+        CGA_LOG_INFO("Started overlap detector");
+        auto overlapper = claragenomics::OverlapperTriggered();
+        auto overlaps = overlapper.get_overlaps(matcher.anchors(), *index);
+
+        CGA_LOG_INFO("Finished overlap detector");
+        std::cerr << "Overlap detection execution time: " << std::chrono::duration_cast<std::chrono::milliseconds>(
+                std::chrono::high_resolution_clock::now() - start_time).count() << "ms" << std::endl;
+
+        overlapper.print_paf(overlaps);
+
+        size_t target_start = query_end + 1;
+        size_t target_end = target_start + index_size;
+
+        if (index.get()->number_of_reads()  < index_size){ //reached the end of the reads
+            break;
+        }
+
+        while(true){ //Now loop over the targets
+            //first generate a2a for query
+            std::vector<std::pair<std::uint64_t, std::uint64_t>> target_ranges;
```

Comment:
We don't strictly know it will only be two ranges, for example we might want an index with only reads over a certain length. I think this way is a little bit more flexible for no performance impact. What do you think?

Reply:
That's what I meant with future-proofing. I'm fine with leaving it as it is for now.